x-pack/plugin/rollup/src/main/java/org/elasticsearch/xpack/rollup/v2/RollupShardIndexer.java

```diff
@@ -279,7 +278,12 @@ public void collect(int docId, long owningBucketOrd) throws IOException {
                         }
 
                         // Create new rollup bucket
-                        rollupBucketBuilder.init(tsid, lastHistoTimestamp);
+                        if (tsidChanged) {
+                            rollupBucketBuilder.init(tsid, lastHistoTimestamp);
+                        } else {
+                            rollupBucketBuilder.init(lastHistoTimestamp);
```

Comment:
I understand the point of minimizing the number of `BytesRef.deepCopyOf()` calls, but I don't like the fact that we "initialize" a bucket without passing the tsid.

Maybe we can handle this in the `RollupBucketBuilder` constructor by

```java
public RollupBucketBuilder init(BytesRef tsid, long timestamp) {
  if (this.tsid == null || this.tsid.equals(tsid) == false) {
    this.tsid = BytesRef.deepCopyOf(tsid);
  }
  ......
}
```

I understand you would still want to avoid the `BytesRef.equals()` at every bucket creation, but I think the performance penalty should be negligible

Reply:
init has two case, one is the tsid changed, the timestamp will also changed. another is timestamp change, it's more common case, if the `BytesRef.equals()` is true ,it will cost more cpu.

I rename the method, and add some comment.